### PR TITLE
`Iris`: Preserve settings on course and exercise update

### PR DESCRIPTION
**Pull request by Hialus**
> [!NOTE]
> Only deploy on TS9

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

Currently there is a bug, where the Iris settings are deleted for a course/programming exercise if they are updated.

### Description

This PR fixes this behaviour by copying the Iris Settings into the new object.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise with Iris enabled (e.g. https://artemis-test9.artemis.cit.tum.de/courses/6/exercises/23)

1. Log in to Artemis
2. Go to the exercise instructor view
3. Edit the exercise -> Iris still enabled
4. Edit the course -> Iris still enabled

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [IrisSettingsService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5488/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.iris/IrisSettingsService.html) | 88% | ✅ |
| [ProgrammingExerciseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5488/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseService.html) | 93% | ✅ |
| [CourseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5488/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/CourseResource.html) | 94% | ✅ |

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/iris/IrisSettingsRepository.java [hunk 1]
```diff
@@ -19,4 +20,14 @@ public interface IrisSettingsRepository extends JpaRepository<IrisSettings, Long
...
+
+    @Query("""
+            SELECT irisSettings
+            FROM IrisSettings irisSettings
+            LEFT JOIN FETCH irisSettings.irisChatSettings ics
+            LEFT JOIN FETCH irisSettings.irisHestiaSettings ihs
+            LEFT JOIN ProgrammingExercise pe ON pe.irisSettings.id = irisSettings.id
+            WHERE pe.id = :#{#programmingExerciseId}
+            """)
+    Optional<IrisSettings> findByProgrammingExerciseId(Long programmingExerciseId);
```
[the remark this answers is not in the document]
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/repository/iris/IrisSettingsRepository.java:32]: Really gotta get this into my head 😅

File: src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java [hunk 2]
```diff
@@ -143,7 +143,7 @@ SELECT CASE WHEN (count(c) > 0) THEN true ELSE false END
     @EntityGraph(type = LOAD, attributePaths = { "lectures", "lectures.lectureUnits" })
     Optional<Course> findWithEagerLecturesAndLectureUnitsById(long courseId);
 
-    @EntityGraph(type = LOAD, attributePaths = { "organizations", "competencies", "prerequisites", "tutorialGroupsConfiguration", "onlineCourseConfiguration" })
+    @EntityGraph(type = LOAD, attributePaths = { "organizations", "competencies", "prerequisites", "tutorialGroupsConfiguration", "onlineCourseConfiguration", "irisSettings" })
     Optional<Course> findWithEagerOrganizationsAndCompetenciesAndOnlineConfigurationById(long courseId);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java:147]: This loads a lot more than the name suggests. Can you maybe rename this to `findCourseForUpdate`?

Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/repository/CourseRepository.java]: Done

**Review by MichaelOwenDyer — DISMISSED** [review]

**Review by JohannesStoehr — DISMISSED** [review]
Code looks good to me

File: src/main/java/de/tum/in/www1/artemis/service/iris/IrisSettingsService.java [hunk 3]
```diff
@@ -375,6 +375,16 @@ public IrisSettings saveIrisSettings(ProgrammingExercise exercise, IrisSettings
...
     }
 
+    /**
+     * Update the Iris settings for a programming exercise.
+     *
+     * @param programmingExerciseBeforeUpdate the programming exercise before the update
+     * @param updatedProgrammingExercise      the programming exercise after the update
+     */
+    public void updateIrisSettings(ProgrammingExercise programmingExerciseBeforeUpdate, ProgrammingExercise updatedProgrammingExercise) {
+        irisSettingsRepository.findByProgrammingExerciseId(programmingExerciseBeforeUpdate.getId()).ifPresent(updatedProgrammingExercise::setIrisSettings);
```
Comment (MaximilianAnzinger) [inline src/main/java/de/tum/in/www1/artemis/service/iris/IrisSettingsService.java:385]: Should there be an error or smth when the settings can't be updated (are not present)?

Or: Add this to the documentation so other developers don't have to look into the source code to figure out how this "edge-case" is handled.

**Review by MaximilianAnzinger — COMMENTED** [review]
Left one suggestion. Other than that code LGTM 👍

File: src/main/java/de/tum/in/www1/artemis/service/iris/IrisSettingsService.java (hunk 3 above)
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/service/iris/IrisSettingsService.java:385]: The settings system will be redone very shortly, so I'll improve it there.

**Pull request merged**